### PR TITLE
Update evaluation hook tests

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Merge tests
- Test arguments

Related to the Lightning hooks review: https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
I like the explicit tests we had before more :)

**Comment by carmocca** [comment]
> I like the explicit tests we had before more :)

I didn't change it to code-golf it. It has the advantage of easily checking any divergence between `validate` and `test`

**Review by tchaton — APPROVED** [review]
Same here. Definitely great value to validate validate vs test, but the structure seems more convoluted now.

File: tests/models/test_hooks.py [hunk 1]
```diff
@@ -314,18 +314,47 @@ def _val_batch():
...
+                dict(name=f'{fn}_step_end', args=({
+                    key: ANY
+                }, )),
+                dict(name=f'Callback.on_{fn}_batch_end', args=(trainer, model, {
+                    key: ANY
+                }, ANY, i, 0)),
+                dict(name=f'on_{fn}_batch_end', args=({
+                    key: ANY
+                }, ANY, i, 0)),
+            ])
```
Comment (Borda) [inline tests/models/test_hooks.py:355]:
```suggestion
        for i in range(batches):
            key_any = {key: ANY}
            out.extend([
                # TODO: `{,Callback}.on_batch_{start,end}`
                dict(name=f'Callback.on_{fn}_batch_start', args=(trainer, model, ANY, i, 0)),
                dict(name=f'on_{fn}_batch_start', args=(ANY, i, 0)),
                dict(name='on_before_batch_transfer', args=(ANY, None)),
                dict(name='transfer_batch_to_device', args=(ANY, torch.device('cpu'), None)),
                dict(name='on_after_batch_transfer', args=(ANY, None)),
                dict(name='forward', args=(ANY, )),
                dict(name=f'{fn}_step', args=(ANY, i)),
                dict(name=f'{fn}_step_end', args=(key_any, )),
                dict(name=f'Callback.on_{fn}_batch_end', args=(trainer, model, key_any, ANY, i, 0)),
                dict(name=f'on_{fn}_batch_end', args=(key_any, ANY, i, 0)),
            ])
```

File: tests/models/test_hooks.py [hunk 2]
```diff
@@ -314,18 +314,42 @@ def _val_batch():
...
+            dict(name=f'on_{fn}_epoch_end'),
+            dict(name='Callback.on_epoch_end', args=(trainer, model)),
+            dict(name='on_epoch_end'),
         ]
 
+    @staticmethod
+    def _eval_batch(fn, trainer, model, batches, key):
+        out = []
+        for i in range(batches):
+            outputs = {key: ANY}
```
Comment (carmocca) [inline tests/models/test_hooks.py:337]:
```suggestion
        out = []
        outputs = {key: ANY}
        for i in range(batches):
```

**Review by Borda — APPROVED** [review]

**Pull request merged**